src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs

```diff
@@ -189,10 +203,16 @@ public MemoryPoolIterator ConsumingStart()
                 {
                     if (!consumed.IsDefault)
                     {
+                        var lengthConsumed = new MemoryPoolIterator(_head).GetLength(consumed);
+
                         returnStart = _head;
                         returnEnd = consumed.Block;
                         _head = consumed.Block;
                         _head.Start = consumed.Index;
+
+                        // Must call Subtract() after bytes have been freed, to avoid producer starting too early and growing
+                        // buffer beyond max length.
```

Comment:
I don't think it matters if you call subtract before or after returning the blocks. I only commented because your comment makes it seem like it does. I think this is just me misconstruing what we mean by the bytes being "freed".